### ISSUE TITLE
Look into options for stricter mandatory access control

### ISSUE THREAD

**Issue by monsieuremre** [secureblue/secureblue#474]
In part replying to @qoijjj in https://github.com/secureblue/secureblue/issues/293. Since the issue is locked, I can not answer you on the original issue.

> Replicating this part and doing it right is the harder part of this feature. If anyone has ideas along these lines, I'm curious to hear them.

A solid way would be a full system MAC policy. First step is to build glibc with hardened malloc only. No weak malloc implementation will be in libc. The default malloc will be a preloadable library compiled separately and not linked. Only select processes would be allowed to preload default malloc. This rule can be enforced using apparmor or selinux.

This still relies on the authenticity of the apps that are exceptions. Since the system is immutable, this might not be a big problem. If every process is confined fully, then this threat might not even be relevant. Changing contexts is extremely restrictive if there is already confinement initially for everything.

It is possible to get a full system apparmor enforcement using an existing project. Fedora does not use apparmor, but it has selinux. Theoretically selinux should be even better since processes themselves can have contexts. But there are no working full system selinux policies for desktop linux distros to my knowledge. The benefits of a MAC policy is of course much more extensive than just enforcing this. So this is something you should consider regardless.

The current state of selinux in fedora is normally not good. It isn't any better than the default apparmor scheme in ubuntu. It might even be worse. In both cases, almost everything is unrestricted except some select services. All real users are unconfined.

Realistically, I cannot feature request that you implement a full system selinux policy. It might have been easier if fedora had apparmor. Selinux is in some ways much better but it is also much difficult. The options that you might consider are to restrict the root user using selinux at least. Consider looking into [this](https://discussion.fedoraproject.org/t/how-to-enable-selinux-user-confinement-what-to-consider/112638/5). This is more realistic, but still requires a lot of work. The other is to work with upstream to make the selinux policy actually confine user processes. There was an incentive once, but I do not know the current situation.

---

You can answer or discuss details or ideas. If this is something you won't look into, close the issue and consider mentioning it in README. Feel free to correct any statements.

**Comment by RoyalOughtness** [comment]
> A solid way would be a full system MAC policy. 

Yep, but this is something that needs to be developed and maintained by a sizeable team, likely upstream.

> except some select services. 

This is underselling it a bit, most all system services are confined.

> The options that you might consider are to restrict the root user using selinux at least. Consider looking into [this](https://discussion.fedoraproject.org/t/how-to-enable-selinux-user-confinement-what-to-consider/112638/5). This is more realistic, but still requires a lot of work.

If you mean confined users in general, then yes I've looked into this. Unfortunately it's far from ready 

> The other is to work with upstream to make the selinux policy actually confine user processes. 

This is far more likely

> You can answer or discuss details or ideas. If this is something you won't look into, close the issue and consider mentioning it in README. Feel free to correct any statements.

It's definitely something to be looked into, so I'll leave this open. But it really needs to happen in the context of what's happening upstream and what tooling is becoming available in the general desktop linux ecosystem. Expanding SELinux into the userspace in particular needs to be done in coordination with upstream to be successful. 

In short, it's something to keep open and collaborate with upstream on, but is a long-term item.

**Comment by RoyalOughtness** [comment]
https://gitweb.gentoo.org/proj/hardened-refpolicy.git/tree/policy/modules/apps/chromium.te
https://gitweb.gentoo.org/proj/hardened-refpolicy.git/tree/policy/modules/apps/chromium.if
https://gitweb.gentoo.org/proj/hardened-refpolicy.git/tree/policy/modules/apps/chromium.fc

**Comment by RoyalOughtness** [comment]
https://github.com/containers/container-selinux/blob/main/container.te#L908

**Comment by monsieuremre** [comment]
I think the approach of disabling unprivileged user name spaces with selinux and then allowing few selected processes manually was a really good decision. It improves the system security by a lot. This shows there are easier steps with selinux that may come before looking into a full system policy.

I think this approach can be extended to limit certain capabilities that have been frequently used in the past to escalate privileges, for everyone including the unrestricted user. Then manually allowing a few processes.

For example limiting the CAP_SETUID and CAP_SETGID with selinux would be more solid than the current systemd service approach of removing bits from binaries. Another very big one to consider would be CAP_PTRACE.

@RoyalOughtness would you look favorable to this?

**Comment by RoyalOughtness** [comment]
https://github.com/secureblue/secureblue/issues/1205
